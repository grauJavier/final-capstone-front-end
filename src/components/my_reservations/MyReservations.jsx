import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import deleteReservation from '../../redux/my_reservations/actions/deleteReservation';
import fetchMyReservations from '../../redux/my_reservations/actions/fetchMyReservations';
import { selectUserId } from '../../redux/user/userSlice.js';
import { formatDistanceToNow } from 'date-fns';

const MyReservations = () => {
  const dispatch = useDispatch();
  const reservations = useSelector((state) => state.myReservations.reservations);
  const userId = useSelector(selectUserId);

  useEffect(() => {
    dispatch(fetchMyReservations(userId));
  }, [dispatch, userId]);

  const handleDelete = (userId, reservationId) => {
    dispatch(deleteReservation({ user_id: userId, reservation_id: reservationId }));
  };

  let flexDirection = String.new;
  const backgroundColors = ['lime-green', 'golden-yellow', 'turquoise-blue', 'charcoal-gray'];

  return (
    <section
      id="my-reservations"
      className="flex items-center justify-center lg:ml-[15%] bg-center py-20"
    >
      <div className="w-[100%] md:px-40 xl:px-80 md:py-20">
        {reservations.length === 0 ? (
          <div className="text-center">
            You haven&apos;t registered any visit yet. <br /> Be brave. Make one.
          </div>
        ) : (
          reservations.map((reservation, index) => {
            if (index % 2 === 0) {
              flexDirection = 'flex-row-reverse';
            } else {
              flexDirection = 'flex-row';
            }

            return (
              <div
                className={`flex ${flexDirection} bg-${backgroundColors[reservation.id % 4]}`}
                key={reservation.id}
              >
                <div
                  className="w-[50%] aspect-square"
                  style={{
                    backgroundImage: `url(${reservation.place.image_url})`,
                    backgroundPosition: 'center',
                    backgroundSize: 'cover',
                  }}
                ></div>
                <div
                  className={`bg-${
                    backgroundColors[index % 4]
                  } dialog-triangle w-[50%] text-white flex flex-col items-center justify-center p-4 gap-2`}
                >
                  <h1 className="text-center secondary-font font-bold text-xl">
                    {reservation.place.name}
                  </h1>
                  <h2 className="text-center">{reservation.place.city.name}</h2>
                  <h3 className="text-center italic">
                    {formatDistanceToNow(new Date(reservation.schedule_date), {
                      addSuffix: true,
                    })}
                  </h3>

                  <div
                    className="cursor-pointer aspect-square p-2 border-white border-2 rounded-full"
                    onClick={() => handleDelete(userId, reservation.id)}
                  >
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      fill="none"
                      viewBox="0 0 24 24"
                      strokeWidth="1.5"
                      stroke="currentColor"
                      className="w-6 h-6"
                    >
                      <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            );
          })
        )}
      </div>
    </section>
  );
}

export default MyReservations;
