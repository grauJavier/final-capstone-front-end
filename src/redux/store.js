import { configureStore, combineReducers } from '@reduxjs/toolkit';
import userReducer from './user/userSlice';
import { persistStore, persistReducer } from 'redux-persist';
import sessionStorage from 'redux-persist/lib/storage/session'; // defaults to localStorage for web
import { createTransform } from 'redux-persist';
import placesSlice from "./places/placesSlice";
import reservationReducer from './reservation_form/reservationSlice.js';
import myReservationsReducer from './my_reservations/myReservationsSlice.js';
import noticeAlertReducer from './notice_alert/noticeAlertSlice.js';

// Create a personalized transform.
const userTransform = createTransform(
  (inboundState) => {
    return {
      currentUser: inboundState.currentUser,
      token: inboundState.token
    };
  },

  (outboundState) => {
    return {
      ...outboundState
    };
  },
  // Apply this transform only to the user state.
  { whitelist: ['user'] }
);

const persistConfig = {
  key: 'root',
  storage: sessionStorage,
  whitelist: ['user'],
  transforms: [userTransform]
};

const reducer = combineReducers({
  noticeAlert: noticeAlertReducer,
  user: userReducer,
  placesSlice,
  reservation: reservationReducer,
  myReservations: myReservationsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }),
});

const persistor = persistStore(store);

export { store, persistor };
