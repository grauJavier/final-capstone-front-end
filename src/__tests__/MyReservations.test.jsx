import { render, screen } from '@testing-library/react';
import MyReservations from '../components/my_reservations/MyReservations';
import { describe, it, expect } from 'vitest';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { store } from '../redux/store';

const mockStore = configureStore([thunk]);

describe('My Reservations Component', () => {
  it('MyReservations renders correctly', () => {
    render(
      <Provider store={store}>
        <MyReservations />
      </Provider>
    );

    expect(MyReservations).toMatchSnapshot();
  });

  it('Should show "You haven\'t registered any visit yet" when there are no reservations', () => {
    render(
      <Provider store={store}>
        <MyReservations />
      </Provider>
    );

    expect(screen.getByText(/You haven't registered any visit yet/)).toBeTruthy();
  });

  it('Should display reservations when there are reservations', () => {
    const fedStore = mockStore({
      user: {
        currentUser: { id: 1, email: 'graujavier@gmail.com', username: 'Javier' },
      },
      myReservations: {
        reservations: [
          {
            id: 1,
            user_id: 1,
            place_id: 1,
            schedule_date: '2023-11-22',
            created_at: '2023-11-08T23:28:39.425Z',
            updated_at: '2023-11-08T23:28:39.425Z',
            place: {
              name: 'Test Hotel',
              image_url: 'https://www.testimage.com/image.jpg',
              description: 'Description of Beautiful Hotel',
              city: {
                name: 'Test City, Test Country',
              },
            },
          },
        ],
      },
    });

    render(
      <Provider store={fedStore}>
        <MyReservations />
      </Provider>
    );

      expect(screen.getByText('Test Hotel')).toBeTruthy();
      expect(screen.getByText('Test City, Test Country')).toBeTruthy();
  });
});
