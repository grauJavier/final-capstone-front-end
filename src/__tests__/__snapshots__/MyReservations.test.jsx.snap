// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`My Reservations Component > MyReservations renders correctly 1`] = `[Function]`;
